server/src/main/java/org/elasticsearch/cluster/coordination/JoinHelper.java

```diff
@@ -232,6 +237,10 @@ void logLastFailedJoinAttempt() {
 
     public void sendJoinRequest(DiscoveryNode destination, long term, Optional<Join> optionalJoin) {
         assert destination.isMasterNode() : "trying to join master-ineligible " + destination;
+        if (nodeHealthService.getHealth() == NodeHealthService.Status.UNHEALTHY) {
+            logger.warn("All paths are not writable. Blocking join request");
```

Comment:
A warning here isn't helpful, we should be logging the failure elsewhere so this will simply result in confusion.

```suggestion
            logger.debug("All paths are not writable. Blocking join request");
```

Also, we have this generic `NodeHealthService` but the log message is very specific: `all paths are not writable`. Maybe the `NodeHealthService` should describe the problem rather than returning a simple boolean.